src/config.js

```diff
@@ -0,0 +1,116 @@
+  const endpoint = env.PLUGIN_ENDPOINT || env.SNOW_ENDPOINT;
+  const newDeployment = /^deployment$/i.test(env.PLUGIN_NOTIFICATION_TYPE || env.SNOW_NOTIFICATION_TYPE);
+  const statusUpdate = /^(status)? *update$/i.test(env.PLUGIN_NOTIFICATION_TYPE || env.SNOW_NOTIFICATION_TYPE);
+  const title = env.PLUGIN_TITLE || env.SNOW_TITLE || `Deployment #${buildNumber} of ${repo}`;
+  const startTime = env.PLUGIN_START_TIME || env.SNOW_START_TIME || moment().format(format);
+  const startMoment = moment(startTime, format);
+  const endTime = env.PLUGIN_END_TIME || env.SNOW_END_TIME || startMoment.add(30, 'minutes').format(format);
+  const descriptionFile = !statusUpdate && loadFromFile(env.PLUGIN_DESCRIPTION_FILE || env.SNOW_DESC_FILE);
+  const description = env.PLUGIN_DESCRIPTION || env.SNOW_DESC || descriptionFile;
+  const testingFile = !statusUpdate && loadFromFile(env.PLUGIN_TESTING_FILE || env.SNOW_TESTING_FILE);
+  const testing = env.PLUGIN_TESTING || env.SNOW_TESTING || testingFile;
+  const commentsFile = !newDeployment && loadFromFile(env.PLUGIN_COMMENTS_FILE || env.SNOW_COMMENTS_FILE);
+  const comments = env.PLUGIN_COMMENTS || env.SNOW_COMMENTS || commentsFile;
+  const success = /^success$/i.test(env.PLUGIN_DEPLOYMENT_OUTCOME || env.SNOW_STATUS || env.DRONE_BUILD_STATUS);
+  const newChange = newDeployment || !(statusUpdate || comments);
+  const failOnError = !/^false$/i.test(env.PLUGIN_FAIL_ON_ERROR || env.SNOW_FAIL_ON_ERROR);
+  const proxy = env.PLUGIN_PROXY || env.SNOW_PROXY;
+  const messageTemplates = {
+    openChange: {
+      messageid: 'HO_SIAM_IN_REST_CHG_POST_JSON',
+      'external_identifier': externalID,
+      payload: {
+        title,
+        startTime,
+        endTime,
+        description: stripControl(description),
+        supplierRef: externalID,
+      }
+    },
+    update: {
+      messageid: 'HO_SIAM_IN_REST_CHG_UPDATE_JSON',
+      'internal_identifier': internalID,
+      payload: {
+        success: success ? 'true' : 'false',
+        comments: stripControl(comments)
+      }
+    }
+  };
+
+  if (moment(endTime, format).isBefore(startMoment)) {
+    throw new RangeError(`The deployment start time MUST be before the end time.
+If you want to set an end time in the past you must also explicitly set a start time before the end time.
+See PLUGIN_START_TIME, SNOW_START_TIME, PLUGIN_END_TIME, SNOW_END_TIME`);
+  }
+
+  if (testing) {
+    messageTemplates.openChange.payload.testing = stripControl(testing);
+  }
+
+  const config = {
+    newChange,
+    endpoint: endpoint || `${protocol}://${sendToProd ? snowProdInstance : snowTestInstance}/${snowPath}`,
+    proxy,
+    username,
+    password,
+    message: newChange ? messageTemplates.openChange : messageTemplates.update,
+    intIDFile,
+    success,
+    failOnError
+  };
+  logger.verbose('config parameters: ' + JSON.stringify(config));
+
+  return config;
+};
+
+module.exports = { configure, loadFromFile, stripControl, DEFAULTS };
```

Comment:
Do all of these need to be exported from the module? Couldn't they go into some sort of config-helpers module?

Reply:
Yes this could be done, but I'd like to do that at a later point as this change is a blocking another I'm already working on. Plus I think a much more useful refactor would be to replace a lot of the `configure` stuff with `yargs`.